File: 543.diameter-of-binary-tree.py
#
# @lc app=leetcode id=543 lang=python3
#
# [543] Diameter of Binary Tree
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from typing import Optional


class Solution:
    def maxDepth(self, node):
        if not node : return 0
        left = self.maxDepth(node.left)
        right = self.maxDepth(node.right)
        self.result = max(self.result, left + right)
        return 1 + max(left, right) 

    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        self.result = 0
        rootMaxDepth = self.maxDepth(root)
        return self.result
        
# @lc code=end

File: 543.diameter-of-binary-tree.cpp
/*
 * @lc app=leetcode id=543 lang=cpp
 *
 * [543] Diameter of Binary Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int result = 0;
    
    int maxDepth(TreeNode* node) {
        if (!node) return 0;
        int left = maxDepth(node->left);
        int right = maxDepth(node->right);
        result = max(result, left + right);
        return 1 + max(left, right);
    }

    int diameterOfBinaryTree(TreeNode* root) {
        result = 0;
        maxDepth(root);
        return result;
    }
};

// @lc code=end

